pkg/controllers/apply_controller.go

```diff
@@ -202,24 +201,27 @@ func (r *ApplyWorkReconciler) applyUnstructured(
 		return nil, false, err
 	}
 
-	if !findOwnerReference(curObj.GetOwnerReferences(), workObj.GetOwnerReferences()[0]) {
+	if !hasSharedOwnerReference(curObj.GetOwnerReferences(), workObj.GetOwnerReferences()[0]) {
 		// TODO: Block All Owner reference in the Work Manifest.
 		err = fmt.Errorf("this object is not owned by the work-api")
 		klog.V(5).InfoS("This object is not owned by the work-api.", "gvr", gvr, "obj", workObj.GetName(), "err", err)
 		return nil, false, err
 	}
 
-	// Compare and update the unstrcuctured.
-	needUpdate := false
-	if !isSameUnstructuredMeta(workObj, curObj) {
+	// Compare the unstructured object and update if needed.
+	updateWarranted := isUpdateWarranted(workObj, curObj)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if updateWarranted {
 		klog.V(5).InfoS("work object meta has changed", "gvr", gvr, "obj", workObj.GetName())
-		needUpdate = true
 		workObj.SetAnnotations(mergeMapOverrideWithDst(curObj.GetAnnotations(), workObj.GetAnnotations()))
 		workObj.SetLabels(mergeMapOverrideWithDst(curObj.GetLabels(), workObj.GetLabels()))
 		workObj.SetOwnerReferences(mergeOwnerReference(curObj.GetOwnerReferences(), workObj.GetOwnerReferences()))
 	}
 
-	if isManifestModified(observedGeneration, curObj, workObj) || needUpdate {
+	if isManifestModified(observedGeneration, curObj, workObj) || updateWarranted {
```

Comment:
why do we still care if the manifest is modified?

Reply:
From what I found this is redundant. Generation is only changed with the spec changes. Our hash annotation takes care of determining if a change has occurred on the manifest spec. Removing this check.